src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -712,6 +712,15 @@ trait Infer extends Checkable {
         argtpes
     }
 
+    // cp. enhanceBounds in typedAppliedTypeTree
+    def enhanceBounds(okparams: List[Symbol], okargs: List[Type], undets: List[Symbol]): Unit =
+      undets.foreach { undet =>
+        val bounds = undet.info.bounds
+        val substBounds = bounds.subst(okparams, okargs)
+        if(bounds ne substBounds)
+          undet.setInfo(substBounds)
```

Comment:
I think we should use `updateInfo` instead of `setInfo`. The latter is problematic here because it wipes out `undet`'s type history. Since we didn't create the `undet` symbol, we can't be sure that's ok.. In practice, this isn't a problem if we're early on in the compiler and there is no type history to wipe out anyway (which is already a tenuous assumption in `typedAppliedTypeTree`). However, type inference definitely can be triggered in later phases.

I seem to remember fixing bugs resulting of similar abuse of setInfo in the GADT arena. It's also possible, as always, that I'm mistaken, but I'd like you to investigate this (e.g., by trying to construct a run where the type history is non-empty and dumping it before calling setInfo), and hear what @lrytz and @retronym think before merging.

Reply:
I'm pretty much convinced that `updateInfo` is the right thing to do rather than `setInfo`, but I've not been able to construct an example with a non-trivial `TypeHistory`. Prior to the `setInfo` call above we have a single entry either for namer or typer, and afterwards the entry is overwritten yielding exactly the same result as if `updateInfo` was called. Clearly this could go wrong if the `TypeHistory` was more interesting, but I can't see a way to get that to happen. I think that might be because these are all undetermined params, and will have been freshly cloned (eg. in `Inferencer.isApplicable`).

I'm happy to follow up on this if anyone can suggest a way of accumulating a more interesting `TypeHistory`, but for now I'm just going to switch to `updateInfo` because that appears to be the right thing to do in any case.

Reply:
The TypeHistory would become more interesting (and relevant) if this method is called later during compilation. Let's say specialization. It's certainly far from trivial to come up with an example (or I would have included it in my comment). If nothing comes up after playing around some more in a debugger and observing what happens, I'm happy to conclude this thread with  the side effect of leaving this breadcrumb in the mind of future contributors :-)